From 43db0f39ab98021fe28870c944a893577e41a8d7 Mon Sep 17 00:00:00 2001
From: Martin Pitt <martin.pitt@ubuntu.com>
Date: Wed, 2 Mar 2011 12:39:41 +0100
Subject: [PATCH] Fix crash when hitting the login button while doing passwordless login

If the user does not need a password to login with gdm, then hitting the "log
in" button causes gdm to crash. This is due to an AnswerQuery DBUS
message being generated even if the password input field is not visible and an
password has not been requested. The user has a real opportunity to hit this
bug at least if pam_mkhomedir is in use and needs some time to finish due to
the size of the skeleton.

Add a has_pending_queries boolean that prevents gdm from issuing a answer query
message if no password has been requested by pam.

https://bugzilla.gnome.org/show_bug.cgi?id=643176
---
 gui/simple-greeter/gdm-greeter-login-window.c |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

diff --git a/gui/simple-greeter/gdm-greeter-login-window.c b/gui/simple-greeter/gdm-greeter-login-window.c
index 6760c8b..f749380 100644
--- a/gui/simple-greeter/gdm-greeter-login-window.c
+++ b/gui/simple-greeter/gdm-greeter-login-window.c
@@ -119,6 +119,7 @@ struct GdmGreeterLoginWindowPrivate
 
         gboolean         user_list_disabled;
         guint            num_queries;
+        gboolean         has_pending_queries;
 
         gboolean         timed_login_already_enabled;
         gboolean         timed_login_enabled;
@@ -353,7 +354,11 @@ on_login_button_clicked_answer_query (GtkButton             *button,
         text = gtk_entry_get_text (GTK_ENTRY (entry));
 
         _gdm_greeter_login_window_set_interactive (login_window, TRUE);
-        g_signal_emit (login_window, signals[QUERY_ANSWER], 0, text);
+
+	if (login_window->priv->has_pending_queries) {
+		login_window->priv->has_pending_queries = FALSE;
+	        g_signal_emit (login_window, signals[QUERY_ANSWER], 0, text);
+	}
 }
 
 static void
@@ -631,6 +636,7 @@ reset_dialog (GdmGreeterLoginWindow *login_window,
         set_sensitive (login_window, FALSE);
 
         login_window->priv->num_queries = 0;
+        login_window->priv->has_pending_queries = FALSE;
 
         if (dialog_mode == MODE_SELECTION) {
                 if (login_window->priv->timed_login_enabled) {
@@ -866,6 +872,7 @@ gdm_greeter_login_window_info_query (GdmGreeterLoginWindow *login_window,
         g_return_val_if_fail (GDM_IS_GREETER_LOGIN_WINDOW (login_window), FALSE);
 
         login_window->priv->num_queries++;
+        login_window->priv->has_pending_queries = TRUE;
         maybe_show_cancel_button (login_window);
 
         g_debug ("GdmGreeterLoginWindow: info query: %s", text);
@@ -898,6 +905,7 @@ gdm_greeter_login_window_secret_info_query (GdmGreeterLoginWindow *login_window,
         g_return_val_if_fail (GDM_IS_GREETER_LOGIN_WINDOW (login_window), FALSE);
 
         login_window->priv->num_queries++;
+        login_window->priv->has_pending_queries = TRUE;
         maybe_show_cancel_button (login_window);
 
         entry = GTK_WIDGET (gtk_builder_get_object (GDM_GREETER_LOGIN_WINDOW (login_window)->priv->builder, "auth-prompt-entry"));
-- 
1.7.4.1

